src/snowflake/snowpark/modin/plugin/extensions/dataframe_overrides.py

```diff
@@ -459,104 +464,222 @@ def __init__(
     # TODO: SNOW-1063346: Modin upgrade - modin.pandas.DataFrame functions
     # Siblings are other dataframes that share the same query compiler. We
     # use this list to update inplace when there is a shallow copy.
-    from snowflake.snowpark.modin.pandas.utils import try_convert_index_to_native
+    from snowflake.snowpark.modin.plugin.extensions.index import Index
 
     self._siblings = []
 
-    # Engine.subscribe(_update_engine)
-    if isinstance(data, (DataFrame, Series)):
-        self._query_compiler = data._query_compiler.copy()
-        if index is not None and any(i not in data.index for i in index):
-            ErrorMessage.not_implemented(
-                "Passing non-existant columns or index values to constructor not"
-                + " yet implemented."
-            )  # pragma: no cover
-        if isinstance(data, Series):
-            # We set the column name if it is not in the provided Series
-            if data.name is None:
-                self.columns = [0] if columns is None else columns
+    # 0. Setting the query compiler
+    # -----------------------------
+    if query_compiler is not None:
+        # CASE I: query_compiler
+        # If a query_compiler is passed in only use the query_compiler field to create a new DataFrame.
+        assert (
+            data is None
+        ), "Invalid DataFrame construction! Cannot pass both data and query_compiler."
+        assert (
+            index is None
+        ), "Invalid DataFrame construction! Cannot pass both index and query_compiler."
+        assert (
+            columns is None
+        ), "Invalid DataFrame construction! Cannot pass both columns and query_compiler."
+        self._query_compiler = query_compiler
+        return
+
+    error_checking_for_init(index, dtype)
+
+    # The logic followed here is:
+    # 1. Create a query_compiler from the provided data. If columns are provided, add/select the columns.
+    # 2. If an index is provided, set the index through set_index or reindex.
+    # 3. If the data is a DataFrame, perform loc to select the required index and columns from the DataFrame.
```

Comment:
you should move all columns operation into Step 1. It is confusing to select columns here again.

Reply:
done!